x-pack/plugins/enterprise_search/public/applications/shared/react_router_helpers/eui_link.tsx

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React from 'react';
+import { useHistory } from 'react-router-dom';
+import { EuiLink, EuiButton, EuiButtonProps, EuiLinkAnchorProps } from '@elastic/eui';
+
+import { letBrowserHandleEvent } from './link_events';
+
+/**
+ * Generates either an EuiLink or EuiButton with a React-Router-ified link
+ *
+ * Based off of EUI's recommendations for handling React Router:
+ * https://github.com/elastic/eui/blob/master/wiki/react-router.md#react-router-51
+ */
+
+interface IEuiReactRouterProps {
+  to: string;
+}
+
+export const EuiReactRouterHelper: React.FC<IEuiReactRouterProps> = ({ to, children }) => {
+  const history = useHistory();
+
+  const onClick = (event: React.MouseEvent) => {
+    if (letBrowserHandleEvent(event)) return;
+
+    // Prevent regular link behavior, which causes a browser refresh.
+    event.preventDefault();
+
+    // Push the route to the history.
+    history.push(to);
```

Comment:
Optional: (I know this means quite a lot of changes as you would need to add a specific context provider instead of using `useHistory`) It would be better to use `applicationSetup.navigateToUrl` here, to handle cases where you'll navigate to an internal link;

https://github.com/elastic/kibana/blob/159369b71999b1f9d8df5645c0935044d25bb8bc/src/core/public/application/types.ts#L742-L769

This can be done as a follow-up.

Reply:
This is great, will definitely look into it as a follow-up PR!